### DISCUSSION TITLE
I think I found a bug

### DISCUSSION THREAD

**Discussion by eiagit** [porsager/postgres#840] [Q&A] [answered]
MY HTTP HEANDLE : `http://127.0.0.1:3000/usuario/pesquisa/?USO_NOME=ENN&USO_CAMPO=USO_NOME`

MY ROUT :  
```js
app.get("/usuario/pesquisa",async(req,res)=>{
    try {
        const params=[];
        params.push(req.query.USO_CAMPO)
        params.push(req.query.USO_NOME)  
         const results = await Execsql  `SELECT "JUSO".*,"TIU_NOME"
        ,"STA_NOME" AS "USO_STANOM"
        ,"USO_FOTO"
        From "JUSO"
        LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
        LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
        WHERE UPPER(${'"'+req.query.USO_CAMPO+'"'}) LIKE 
        ${String.fromCharCode(39)+'%'+req.query.USO_NOME+'%'+String.fromCharCode(39)}
       // ${"'%"+req.query.USO_NOME+"%'"}
        ORDER BY "USO_NOME"`
        res.status(200).json(results.USO_ID)
        console.log(results)
    } catch (err) {
        console.log(err)
        res.status(500).json({ 'error': err })
    }  
```

THE RESULTS IS `[]`
but when I use a result of my script : 

```sql
SELECT "JUSO".*,"TIU_NOME"
        ,"STA_NOME" AS "USO_STANOM"
        ,"USO_FOTO"
        From "JUSO"
        LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
        LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
        WHERE UPPER("USO_NOME") LIKE '%ENN%'
        ORDER BY "USO_NOME"
```
it returns my data, but both scripts are the same 

any one can ansori this mistery ?

thank you

**Comment by porsager** [comment]
Can you try to fix the formatting so that your code becomes readable?

**Reply by eiagit** [reply]
ok, I try write more readable .
```javascript
try {
const results = await Execsql`SELECT "JUSO".*
,"TIU_NOME"
 ,"STA_NOME" AS "USO_STANOM" 
,"USO_FOTO" 
From "JUSO" LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO" 
LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS" 
WHERE UPPER(${'"'+req.query.USO_CAMPO+'"'}) 
LIKE  ${"'%"+req.query.USO_NOME+"%'"} 
res.status(200).json(results.USO_ID)
} catch (err) {
res.status(500).json({ 'error': err })
}
``` 


I pass this string
bad this return [ ] 
when a pass direct this string then result are my data : 

```javascript
const results = await Execsql`SELECT "JUSO".*,"TIU_NOME"
,"STA_NOME" AS "USO_STANOM"
,"USO_FOTO"
From "JUSO"
LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
WHERE UPPER("USO_NOME") LIKE '%ENN%'
ORDER BY "USO_NOME"`
``` 


cant read ? 

MY HTTP HEANDLE : http://127.0.0.1:3000/usuario/pesquisa/?USO_NOME=ENN&USO_CAMPO=USO_NOME

the codes are the same, one is dynamic and the other I typed just for testing, and the typed one brings the results and the dynamic one brings nothing.

I wouldn't want to be here trying to solve a possible bug, I'm using it exactly as it is in the documentation.

If you want to see the error, we can arrange an appointment via AnyDesk

I've tried everything, spent hours tinkering with the code, changing it and the result is always the same.

the fault is here: 

LIKE  ${''%"+req.query.USO_NOME+"%'"} 

try to any javascript code ok
than you

**Reply by porsager** [reply]
You can edit your original post, and then read up on markdown formatting to make it readable using code blocks (` ``` `) ☺️ No need to post again 😉

**Reply by eiagit** [reply]
my AnyDesk 1535876956 
you see fast

**Reply by porsager** [reply]
It sounds like you're looking for paid support. This is an open source product, there is not the kind of support you are requesting. I don't have time to do paid support at the moment, so either clean up your issue and do the bare minimum which is expected to be lucky enough to receive free help, or I'll delete it.

**Reply by eiagit** [reply]
this no are supoert is a contribuiton

**Reply by porsager** [reply]
No it is not, you clearly have an obvious error in your code - go read the documentation

**Reply by eiagit** [reply]
ok, thank you, I belive in you, but try your self :) you are see.

**Reply by porsager** [reply]
It's so obvious you didn't read the documentation as you are trying in weird ways to put unsupported quotes around the string in your arguments. Even though the code you posted is not very readable what you're doing doesn't make any sense at all. Had you done the minimum to format your message instead of continuing your entitled approach you might have gotten a better response. Please try to be respectful of other peoples time in the future.

**Reply by eiagit** [reply]
the document : 
```javascript
const users = await sql`
  select
    name,
    age
  from users
  where
    name like ${ name + '%' }
    and age > ${ age }
``` 
my code 👍
```javascript
LIKE  ${"'%"+req.query.USO_NOME+"%'"} 
``` 

I try 
```javascript
   ${'%'+req.query.USO_NOME+'%'}
``` 


It's not my first clash with a developer. and I don't think it will be the last.
respect is always good in two ways.
stay cool, there is a fault, I hope it is mine, from the bottom of my heart.
I spent the whole day yesterday changing this script, before sending this message to you.

**Reply by porsager** [reply]
If you don't change your attitude, then no, you will continue to have clashes, and you will drain the energy from open source developers eventually being one of the causes that open source might die. Remember that a problem on your end doesn't result in an obligation on my end. Even if there should be an error in the code I have made available freely, you are not entitled to anything. You have been given a gift, and if you choose to receive it or use it there is no obligation on the part of the gift giver. If you don't like it, don't use it.

Be grateful that I'm trying to open your eyes instead of simply deleting your issue, and that I am now even trying to help despite your bad start.

Now your report here is a bit better, and I trust that you've actually tried. The last example you posted (`${ '%' + req.query.USO_NOME + '%' }` ) is the way it should be done if you want to find rows where the name contains the value of `req.query.USO_NAME`, so you should take care to try and see if the value is actually what you think it is. Try to do `console.log('%'+req.query.USO_NOME+'%')` above your query to see if it makes any sense.

**Reply by eiagit** [reply]
sorry, that's not my intention, after a day of trying, my stress is high.
I understand that you have no obligation to help me. one more time, sorry.
because I write to you:
I used this code with mysql2 and ok, I used this code with node-firebird and ok.
but using postgres 3.4.4 without success.

I've been programming for a long time. for more exactly 45 years, basic, turbo, basic, pascal, delphi, phyton and others. Now Javascript.

I'm a fan of free software, I write articles and  codes to make life easier for beginners.

Because my code is weird: it's dynamic.
column and data are passed by the user and this makes it a bit strange.

I will continue trying until I achieve a dynamic query, regarding columns and data until I reach the conclusion that this library does not support what I want to do.
Once again, a thousand apologies, words are cold and don't convey feelings.
Under no circumstances did I want to disrespect you or diminish your contribution to humanity, just to collaborate.
If this interface supports photos, I'll send some, maybe that will help me understand better. Grateful

**Reply by porsager** [reply]
Thank you.

If you can send the code properly formatted, it will be easier to spot if there are any obvious mistakes. Just insert the code between 2 lines that contain 3 backticks like this:

```` 
```
code here
```
````

**Reply by eiagit** [reply]
![image](https://github.com/porsager/postgres/assets/43192255/60a802d8-d5d2-42af-bd88-de0d75c86167)
this is with I waited 

![image](https://github.com/porsager/postgres/assets/43192255/91e136af-c136-46a1-9df6-345cf6c77482)
this is with I see my code

![image](https://github.com/porsager/postgres/assets/43192255/38464e9d-eb46-4f91-9b7b-d7d6f9fafb0c)
the result of console.log

![image](https://github.com/porsager/postgres/assets/43192255/72173cf6-a544-4fce-8ee6-57dd4b72fecf)
the code

![image](https://github.com/porsager/postgres/assets/43192255/6fcc2d99-4299-41d2-b2bf-f7c83fde81f9)
the result : [ ]

That's what I see, do you understand why I wrote to you? , I took the result from my console.log , put it from sql and the result was ok. but, when I try to go through "postgres 3..4.4" it doesn't bring any results.
I'll keep insisting until I succeed :) .

**Reply by porsager** [reply]
You need to change
```
WHERE ${'"'+req.query.USO_CAMPO+'"'} LIKE
${"'%'+req.query.USO_NOME+"%'"}
```

to

```
WHERE ${ sql(req.query.USO_CAMPO) } LIKE
${ '%' + req.query.USO_NOME +  '%' }
```

But be aware that this will allow users of your api to search any column.

Besides from that you should really consider using snake case for your postgres columns, eg (`uso_campo`) rather than `USO_CAMPO`. That will result in not having to use quotes everywhere.

**Reply by eiagit** [reply]
```javascript 
app.get("/usuario/pesquisa",async(req,res)=>{
    try {
        const results = await Execsql  `SELECT "JUSO".*,"TIU_NOME"
        ,"STA_NOME" AS "USO_STANOM"
        ,"USO_FOTO"
        From "JUSO"
        LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
        LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
        WHERE ${'"'+req.query.USO_CAMPO+'"'} LIKE 
        ${"'%"+req.query.USO_NOME+"%'"} 
        ORDER BY "USO_NOME"`
        res.status(200).json(results)
        console.log(results)
    } catch (err) {
        console.log(err)
        res.status(500).json({ 'error': err })
    }         
})
```

with others wrote this code, but it would not be dynamic  column and data :
```javascript
SELECT "JUSO".*,"TIU_NOME"
        ,"STA_NOME" AS "USO_STANOM"
        ,"USO_FOTO"
        From "JUSO"
        LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
        LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
        WHERE "USO_NOME" LIKE
        '%ENN%'
        ORDER BY "USO_NOME"
```

**Reply by porsager** [reply]
Compare this:
```sql
SELECT "JUSO".*,"TIU_NOME"
,"STA_NOME" AS "USO_STANOM"
,"USO_FOTO"
From "JUSO"
LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
WHERE UPPER("USO_NOME") LIKE '%ENN%'
ORDER BY "USO_NOME"
```

with this
```sql
select 
  juso.*, 
  tiu_nome,
  sta_nome as uso_stanom,
  uso_foto
from juso
left join jtiu on tiu_id = uso_tipo
left join jsta on sta_id = uso_status
where upper(uso_nome) like '%ENN%'
order by uso_nome
```

**Reply by porsager** [reply, marked as answer]
Ok, this should work with your current names
```js
app.get("/usuario/pesquisa",async(req,res)=>{
    try {
        const results = await Execsql`
          SELECT "JUSO".*,"TIU_NOME"
          ,"STA_NOME" AS "USO_STANOM"
          ,"USO_FOTO"
          From "JUSO"
          LEFT JOIN "JTIU" ON "TIU_ID"="USO_TIPO"
          LEFT JOIN "JSTA" ON "STA_ID"="USO_STATUS"
          WHERE ${ sql(req.query.USO_CAMPO) } LIKE 
          ${ '%' + req.query.USO_NOME + '%' } 
          ORDER BY "USO_NOME"
        `
        res.status(200).json(results)
        console.log(results)
    } catch (err) {
        console.log(err)
        res.status(500).json({ 'error': err })
    }         
})
```

but this is much more manageable if you change your naming and do a little prettier formatting

```js
app.get('/usuario/pesquisa', async (req, res) => {
  try {
    const results = await sql`
      select 
        juso.*,
        tiu_nome,
        sta_nome as uso_stanom
        uso_foto
      from juso
      left join jtiu on tiu_id = uso_tipo
      left join jsta on sta_id = uso_status
      where ${ sql(req.query.uso_campo) } like ${ '%' + req.query.uso_nome + '%' } 
      order by uso_nome
    `
    res.status(200).json(results)
    console.log(results)
  } catch (err) {
    console.log(err)
    res.status(500).json({ 'error': err })
  }         
})
```

**Comment by eiagit** [comment]
you are rason , the problem is with i write my code. 
with your changes the result are ok.
thank you  of your suport.
1000 sorryes.

**Reply by porsager** [reply]
Great to hear it's working for you now 👍 Good luck with your project

**Comment by eiagit** [comment]
this is only a test to connect any database. than you and Good luck.
